src/typeChecker/typeErrorChecker.ts

```diff
@@ -0,0 +1,914 @@
+      const value = typeNode.literal.value
+      if (typeof value !== 'string' && typeof value !== 'number' && typeof value !== 'boolean') {
+        throw new TypecheckError(typeNode, 'Unknown literal type')
+      }
+      return tLiteral(value)
+    case 'TSIntersectionType':
+      throw new TypecheckError(typeNode, 'Intersection types are not allowed')
+    case 'TSTypeReference':
+      const declaredType = lookupTypeAlias(typeNode.typeName.name, env)
+      if (!declaredType) {
+        context.errors.push(new TypeNotFoundError(typeNode, typeNode.typeName.name))
+        return tAny
+      }
+      return declaredType
+    default:
+      return getPrimitiveType(typeNode, context)
+  }
+}
+
+/**
+ * Converts array of function parameters into array of types.
+ */
+function getParamTypes(params: tsEs.Identifier[], context: Context, env: TypeEnvironment): Type[] {
+  return params.map(param => getTypeAnnotationType(param.typeAnnotation, context, env))
+}
+
+/**
+ * Converts node type to primitive type, adding errors to context if disallowed/unknown types are used.
+ * If errors are found, returns the "any" type to prevent throwing of further errors.
+ */
+function getPrimitiveType(node: tsEs.TypeKeywordNode, context: Context) {
+  const primitiveType = typeAnnotationKeywordToBasicTypeMap[node.type] ?? 'unknown'
+  if (
+    disallowedTypes.includes(primitiveType as TSDisallowedTypes) ||
+    (context.chapter === 1 && primitiveType === 'null')
+  ) {
+    context.errors.push(new TypeNotAllowedError(node, primitiveType))
+    return tAny
+  }
+  return tPrimitive(primitiveType as PrimitiveType | TSAllowedTypes)
+}
+
+/**
+ * Wrapper function for lookupType that removes ForAll and Predicate types,
+ * since they are not used in the type error checker.
+ */
+function lookupTypeAndRemoveForAllAndPredicateTypes(
+  name: string,
+  env: TypeEnvironment
+): Type | undefined {
+  const type = lookupType(name, env)
+  if (!type) {
+    return undefined
+  }
+  if (type.kind === 'forall') {
+    // Skip typecheck as function has variable number of arguments;
+    // this only occurs for certain prelude functions
+    return tAny
+  }
+  if (type.kind === 'predicate') {
+    // All predicate functions (e.g. is_number)
+    // take in 1 parameter and return a boolean
+    return tFunc(tAny, tBool)
+  }
+  return type
+}
+
+/**
+ * Combines all types provided in the parameters into one, removing duplicate types in the process.
+ */
+function mergeTypes(...types: Type[]): Type {
+  const mergedTypes: Type[] = []
+  for (const currType of types) {
+    if (isEqual(currType, tAny)) {
+      return tAny
+    }
+    if (currType.kind === 'union') {
+      for (const type of currType.types) {
+        if (!containsType(mergedTypes, type)) {
+          mergedTypes.push(type)
+        }
+      }
+    } else if (!containsType(mergedTypes, currType)) {
+      mergedTypes.push(currType)
+    } else {
+      // Duplicate type, do nothing
+    }
```

Comment:
```suggestion
    } else {
      if (!containsType(mergedTypes, currType)) {
        mergedTypes.push(currType)
      }
    }
```

Minor nit: Control flow might be easier to understand if each level does only one thing. Here, the first level is checking the `kind` of `currType`, while the second level is checking whether to push the type into `mergedTypes`.